### DISCUSSION TITLE
NuGet Private Feed (Github packages) Doesn't Authenticate

### DISCUSSION THREAD

**Discussion by 1kevgriff** [renovatebot/renovate#19286] [Request Help] [answered]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

We're having issues setting up Renovate to look at our private NuGet feed hosted on Github Packages.

My renovate.json:

```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base"
  ],
  "nuget": {
    "registryUrls": [
      "https://api.nuget.org/v3/index.json",
      "https://nuget.pkg.github.com/<OUR ORG>/index.json"
    ]
  },
  "hostRules": [
    {
      "hostType": "nuget",
      "matchHost": "https://nuget.pkg.github.com/<OUR ORG>/index.json",
      "username": "<Username Hidden>",
      "password": "<PAT Hidden>"
    }
  ]
}
```

We've tested username and PAT in this case, and they work (so permissions are good).

However when revonate tries to look at the NuGet package, it fails:

```
DEBUG: GET https://nuget.pkg.github.com/<OUR ORG>/<GOOD PACKAGE>/index.json = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=218)
DEBUG: Datasource unauthorized
{
  "datasource": "nuget",
  "packageName": "<GOOD PACKAGE>",
  "url": "https://nuget.pkg.github.com/<OUR ORG>/<GOOD PACKAGE>/index.json"
}
DEBUG: Failed to look up nuget dependency <GOOD PACKAGE> (<GOOD PACKAGE>)(packageFile="<SOURCE PROJECT>.csproj", dependency="<GOOD PACKAGE>")
```

### Relevant debug logs

<details><summary>Logs</summary>

```
DEBUG: GET https://nuget.pkg.github.com/<OUR ORG>/<GOOD PACKAGE>/index.json = (code=ERR_NON_2XX_3XX_RESPONSE, statusCode=401 retryCount=0, duration=218)
DEBUG: Datasource unauthorized
{
  "datasource": "nuget",
  "packageName": "<GOOD PACKAGE>",
  "url": "https://nuget.pkg.github.com/<OUR ORG>/<GOOD PACKAGE>/index.json"
}
DEBUG: Failed to look up nuget dependency <GOOD PACKAGE> (<GOOD PACKAGE>)(packageFile="<SOURCE PROJECT>.csproj", dependency="<GOOD PACKAGE>")
```

</details>

### Have you created a minimal reproduction repository?

No reproduction repository

**Comment by viceice** [comment, marked as answer]
Your host rule needs to be like:
```json
{
  "$schema": "https://docs.renovatebot.com/renovate-schema.json",
  "extends": [
    "config:base"
  ],
  "nuget": {
    "registryUrls": [
      "https://api.nuget.org/v3/index.json",
      "https://nuget.pkg.github.com/<OUR ORG>/index.json"
    ]
  },
  "hostRules": [
    {
      "hostType": "nuget",
      "matchHost": "https://nuget.pkg.github.com/<OUR ORG>/",
      "username": "<Username Hidden>",
      "password": "<PAT Hidden>"
    }
  ]
}
```

this is required to also match `https://nuget.pkg.github.com/<OUR ORG>/<GOOD PACKAGE>/index.json` (prefix match)

**Reply by 1kevgriff** [reply]
Thanks!  sorry, I must have misunderstood entirely the docs around `hostRules`.  But your answer makes complete sense.

**Reply by viceice** [reply]
don't we have samples inside the docs for GitHub packages?

**Reply by 1kevgriff** [reply]
Not directly.  My confusion was on https://docs.renovatebot.com/nuget/#authenticated-feeds

```json
{
  "hostRules": [
    {
      "hostType": "nuget",
      "matchHost": "http://example1.com/nuget",
      "username": "root",
      "password": "p4$$w0rd"
    }
  ]
}
```

I thought `matchHost` was a direct check for whatever was in 

```json
{
  "nuget": {
    "registryUrls": [
      "https://api.nuget.org/v3/index.json",
      "https://example1.com/nuget/",
      "https://example2.com/nuget/v3/index.json"
    ]
  }
}
```

To back it up, if I look at `matchHost` on https://docs.renovatebot.com/configuration-options/#matchhost is has a note

> The matchHost URL must be the same as the registryUrl set in .npmrc

Again made me assume the URLs should be a full match.  Does that make sense?

**Reply by 1kevgriff** [reply]
But your tweak worked 💯 for me, though.
